core/services/relay/types/relay.go

```diff
@@ -13,32 +13,55 @@ import (
 type Network string
 
 var (
-	EVM    Network = "evm"
-	Solana Network = "solana"
-	Terra  Network = "terra"
+	EVM             Network = "evm"
+	Solana          Network = "solana"
+	Terra           Network = "terra"
+	SupportedRelays         = map[Network]struct{}{
+		EVM:    {},
+		Solana: {},
+		Terra:  {},
+	}
 )
 
-// RelayerCtx represents a relayer
-type RelayerCtx interface {
+// PluginArgs are the args required to create any OCR2 plugin components.
+// Its possible that the plugin config might actually be different
+// per relay type, so we pass the config directly through.
+type PluginArgs struct {
+	ConfigWatcherArgs
+	TransmitterID string
+	PluginConfig  []byte
+}
+
+type ConfigWatcherArgs struct {
+	ExternalJobID uuid.UUID
+	JobID         int32
+	ContractID    string
+	RelayConfig   []byte
+}
+
+type Relayer interface {
 	services.ServiceCtx
-	// NewOCR2Provider is generic for all OCR2 plugins on the given chain.
-	NewOCR2Provider(externalJobID uuid.UUID, spec interface{}) (OCR2ProviderCtx, error)
-	// TODO: Will need some CCIP plugin providers for chain specific implementations
-	// of request reading and tracking report status on dest chain.
-	// For now, the ocr2/plugins/ccip is EVM specific.
+	NewConfigWatcher(args ConfigWatcherArgs) (ConfigWatcher, error)
+	NewMedianProvider(args PluginArgs) (MedianProvider, error)
```

Comment:
These are both services (`ConfigWatcher` & `MedianProvider`), services that provide some hooks/ports that can be used by whoever is running those services (OCR2).

They have different naming conventions jet they both provide hooks.
They have different args naming conventions, yet they are both "plugins". Or better yet, services that provide components to build a specific plugin.

Maybe it's a nitpick, but IMO we should tweak the naming here to make the code more readable and understandable.

Reply:
Additionally,  one extends the other (`MedianProvider` is `ConfigWatcher`) but it's not exactly clear they are used for different use cases of the same job. Former to run the oracle job, latter to run the bootstrap job.

Reply:
```go
type Plugin interface {
	services.ServiceCtx
	NewConfigProvider(args ConfigArgs) (ConfigProvider, error)
	NewMedianProvider(args MedianArgs) (MedianProvider, error)
}
```

Should we also namespace the OCR2 protocol?
This way we could namespace different protocols and versions.

Here is a quick example how this could read:

```go
solana := relay.Plugins(relay.Solana)
bootstrapProvider := solana.OCR2.NewConfigProvider(ocr2.ConfigArgs{})
oracleProvider := solana.OCR2.NewMedianProvider(ocr2.MedianArgs{})
```

Reply:
yeah fair enough, maybe just ConfigProvider and MedianProvider is cleaner?